### PR TITLE
feat(core): General purpose `DescriptionAuthorizer`

### PR DESCRIPTION
**Pull request by ajordens**
Ported the `authorize()` method from `DescriptionValidator` and added
a metric in the event that a description presents an account _but_ no
application(s).

This metric will emit some false positives but still prove useful in the
short-term to highlight descriptions that might not implement
`ApplicationNameable` (or not implement it properly).

The previous behavior expected a `Description` to implement
`DescriptionValidator` in order to have authorization applied, however
this is not a requirement and missed some legitimate scenarios.

### CODE REVIEW

File: clouddriver-core/src/main/groovy/com/netflix/spinnaker/clouddriver/deploy/DescriptionAuthorizer.java [hunk 1]
```diff
@@ -0,0 +1,96 @@
...
+      applications.forEach(application -> {
+        if (!fiatPermissionEvaluator.hasPermission(auth, application, "APPLICATION", "WRITE")) {
+          errors.reject("authorization", format("Access denied to application %s", application));
+        }
+      });
+    }
+
+
+    if (account != null && applications.isEmpty()) {
+      registry
```
Comment (ajordens) [inline clouddriver-core/src/main/groovy/com/netflix/spinnaker/clouddriver/deploy/DescriptionAuthorizer.java]: Will also log a suitable message ... as mentioned in the PR description I'm looking to get an idea of how many descriptions are account restricted but not application-restricted.

Likely also worth a `success` tag on this counter.

The plan would be to mark the descriptions that legitimately have no applications, and fix the others.

File: clouddriver-core/src/main/groovy/com/netflix/spinnaker/clouddriver/deploy/DescriptionValidator.groovy [hunk 2]
```diff
@@ -33,40 +33,4 @@ public abstract class DescriptionValidator<T> implements VersionedCloudProviderO
   }
 
   abstract void validate(List priorDescriptions, T description, Errors errors)
-
-  @Autowired(required = false)
-  FiatPermissionEvaluator permissionEvaluator
-
-  void authorize(T description, Errors errors) {
```
Comment (ajordens) [inline clouddriver-core/src/main/groovy/com/netflix/spinnaker/clouddriver/deploy/DescriptionValidator.groovy]: Discovered there are a handful (or more!) descriptions that do not implement `DescriptionValidator` and thus are ignored as far as authorization checks go.

I don't think `implements DescriptionValidator` should be a pre-req and this logic has been extracted / converted to Java.

**Comment by ajordens** [comment]
I have some actual testing to do but wanted to make sure nothing looked off with this general strategy @ttomsu?

I've got another pending PR that touches all the aws and titus descriptions and makes them
 implement the *Nameable interfaces as appropriate. 

Would like to get this change merged prior tho.

File: clouddriver-web/src/main/groovy/com/netflix/spinnaker/clouddriver/controllers/OperationsController.groovy [hunk 3]
```diff
@@ -161,24 +164,18 @@ class OperationsController {
         )
         if (validator) {
           validator.validate(descriptions, description, errors)
-          validator.authorize(description, errors)
         } else {
-          def msg = "No validator found for operation `${description?.class?.simpleName}` and cloud provider $cloudProvider"
-
-          switch(opsSecurityConfigProps.onMissingValidator) {
-            case SecurityConfig.SecurityAction.WARN:
```
Comment (ajordens) [inline clouddriver-web/src/main/groovy/com/netflix/spinnaker/clouddriver/controllers/OperationsController.groovy]: This seems unnecessary now that validation != authorization.

**Comment by ttomsu** [comment]
I like the change.

I made a half-hearted attempt to build in some automatic checking of this stuff: https://github.com/spinnaker/clouddriver/blob/29c5c1435ab0c750c4b491208f600647a3e98901/clouddriver-web/src/main/groovy/com/netflix/spinnaker/clouddriver/listeners/OperationsTypeChecker.groovy#L64

Not sure if you found that or would find it useful to ferret out additional non-compliant Descriptions.

**Pull request merged**